Framework/MDAlgorithms/src/ReplicateMD.cpp

```diff
@@ -293,8 +293,25 @@ void ReplicateMD::exec() {
    the linear index -> linear index calculation below will not work correctly.
    */
   MDHistoWorkspace_const_sptr transposedDataWS = dataWS;
-  if (dataWS->getNumDims() == shapeWS->getNumDims()) {
+  if (nDimsData <= nDimsShape) {
     auto axes = findAxes(*shapeWS, *dataWS);
+    // Check that the indices stored in axes are compatible with the
+    // dimensionality of the data workspace
+    const auto numberOfDimensionsOfDataWorkspace = static_cast<int>(nDimsData);
+    int axisDimCounter = 0;
+    for (const auto &axis : axes) {
+      if (axis >= numberOfDimensionsOfDataWorkspace) {
+        std::string message =
+            "ReplicateMD: Cannot transpose the data workspace. Attempting to "
+            "swap dimension index " +
+            std::to_string(axisDimCounter) + " with index " +
```

Comment:
Can you obtain axisDimCounter with `&axis - axes.data()` or `std::distance(&axis,axes.data())` and remove the counter?